test/cbmc/sources/core_json_contracts.c

```diff
@@ -0,0 +1,185 @@
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef CORE_JSON_CONTRACTS_C_
+#define CORE_JSON_CONTRACTS_C_
+
+#include "core_json_contracts.h"
+
+bool JSON_SearchConstPreconditions( char * buf,
+                                    size_t max,
+                                    char * query,
+                                    size_t queryLength,
+                                    char ** outValue,
+                                    size_t * outValueLength,
+                                    JSONTypes_t * outType )
+{
+    return ( max < CBMC_MAX_BUFSIZE )
+           & ( queryLength < CBMC_MAX_QUERYKEYLENGTH )
+           & ( buf == NULL || allocated( buf, max ) )
+           & ( query == NULL || allocated( query, queryLength ) )
+           & ( outValue == NULL || allocated( outValue, sizeof( *outValue ) ) )
+           & ( outValueLength == NULL || allocated( outValueLength, sizeof( *outValueLength ) ) )
+           & ( outType == NULL || allocated( outType, sizeof( *outType ) ) );
+}
+
+bool JSON_SearchConstPostconditions( JSONStatus_t result,
+                                     char * buf,
+                                     char ** outValue,
+                                     size_t * outValueLength,
+                                     size_t max )
+{
+    CHECK( isJSONSearchEnum( result ) );
+
+    if( result == JSONSuccess )
+    {
+        char * endOfValue = *outValue + *outValueLength;
+        char * endOfBuf = buf + max;
+        return pointer_in_range( buf, endOfValue, endOfBuf );
+    }
+
+    return true;
+}
+
+bool JSON_IteratePreconditions( char * buf,
+                                size_t max,
+                                size_t * start,
+                                size_t * next,
+                                JSONPair_t * outPair )
+{
+    return ( 0 < max && max < CBMC_MAX_BUFSIZE )
+           & ( buf == NULL || allocated( buf, max ) )
+           & ( start == NULL || allocated( start, sizeof( *start ) ) )
+           & ( next == NULL || allocated( next, sizeof( *next ) ) )
+           & ( outPair == NULL || allocated( outPair, sizeof( *outPair ) ) )
+           /* *INDENT-OFF* */
+           & ( outPair != NULL ==> ( ( outPair->keyLength == 0 && outPair->key == NULL ) || allocated( outPair->key, outPair->keyLength ) ) )
+           & ( outPair != NULL ==> ( ( outPair->valueLength == 0 && outPair->value == NULL ) || allocated( outPair->value, outPair->valueLength ) ) );
+           /* *INDENT-ON* */
+}
+
+JSONStatus_t JSON_ValidatePreconditions( char * buf,
+                                         size_t max )
+{
+    return ( max < CBMC_MAX_BUFSIZE )
+           & ( buf == NULL || allocated( buf, max ) );
+}
+
```

Comment:
Can we move that after `JSON_IteratePostconditions` so that `JSON_IteratePreconditions` and `JSON_IteratePostconditions` are next to each other ?